kernel/rtlil.h

```diff
@@ -673,19 +687,68 @@ struct RTLIL::Const
 	bool operator ==(const RTLIL::Const &other) const;
 	bool operator !=(const RTLIL::Const &other) const;
 
+	bitvectype& assert_get_bits(const char* ctx) const;
+	std::string& assert_get_str(const char* ctx) const;
+
+	const std::vector<RTLIL::State>& bits() const;
+	std::vector<RTLIL::State>& bits();
 	bool as_bool() const;
 	int as_int(bool is_signed = false) const;
-	std::string as_string() const;
+	std::string as_string(std::string any = "-") const;
 	static Const from_string(const std::string &str);
+	std::vector<RTLIL::State> to_bits() const;
+	std::string pretty_fmt() const;
+	std::string pretty_fmt_undef() const;
+
 
 	std::string decode_string() const;
+	inline size_t size() const {
+		if (auto str = std::get_if<std::string>(&backing))
+			return 8 * str->size();
+		else
+			return assert_get_bits("Const::size").size();
+	}
+
+	inline bool empty() const {
+		if (auto str = std::get_if<std::string>(&backing))
+			return str->empty();
+		else
+			return assert_get_bits("Const::empty").empty();
+	}
+
+	void bitvectorize() const {
+		if (std::get_if<bitvectype>(&backing))
+			return;
+
+		std::string& str = assert_get_str("Const::bitvectorize");
+		bitvectype bits;
+		bits.reserve(str.size() * 8);
+		for (int i = str.size() - 1; i >= 0; i--) {
+			unsigned char ch = str[i];
+			for (int j = 0; j < 8; j++) {
+				bits.push_back((ch & 1) != 0 ? State::S1 : State::S0);
+				ch = ch >> 1;
+			}
+		}
+		backing = bits;
+	}
 
-	inline int size() const { return bits.size(); }
-	inline bool empty() const { return bits.empty(); }
-	inline RTLIL::State &operator[](int index) { return bits.at(index); }
-	inline const RTLIL::State &operator[](int index) const { return bits.at(index); }
-	inline decltype(bits)::iterator begin() { return bits.begin(); }
-	inline decltype(bits)::iterator end() { return bits.end(); }
+	inline RTLIL::State &operator[](int index) {
+		bitvectorize();
+		return assert_get_bits("Const::operator[]").at(index);
+	}
+	inline const RTLIL::State &operator[](int index) const {
+		bitvectorize();
+		return assert_get_bits("const Const::operator[]").at(index);
+	}
+	inline bitvectype::iterator begin() {
+		bitvectorize();
+		return assert_get_bits("Const bit iterator begin()").begin();
```

Comment:
I think this is not constant, you can make changes through this iterator

Reply:
Yeah, I meant to refer to `RTLIL::Const`, not the C++ `const` keyword

Reply:
OK, then we may drop Const from there as it's superfluous

Reply:
This wrapper is removed and we lean on exceptions in the unexpected case of variant access problems instead